drivers/serial/uart_stm32.c

```diff
@@ -1687,6 +1688,73 @@ static int uart_stm32_init(const struct device *dev)
 #endif
 }
 
+#ifdef CONFIG_PM_DEVICE
+static void uart_stm32_suspend_setup(const struct device *dev)
+{
+	const struct uart_stm32_config *config = dev->config;
+
+#ifdef USART_ISR_BUSY
+	/* Make sure that no USART transfer is on-going */
+	while (LL_USART_IsActiveFlag_BUSY(config->usart) == 1) {
+	}
+#endif
+	while (LL_USART_IsActiveFlag_TC(config->usart) == 0) {
+	}
+#ifdef USART_ISR_REACK
+	/* Make sure that USART is ready for reception */
+	while (LL_USART_IsActiveFlag_REACK(config->usart) == 0) {
+	}
+#endif
+	/* Clear OVERRUN flag */
+	LL_USART_ClearFlag_ORE(config->usart);
+}
+
+static int uart_stm32_pm_action(const struct device *dev,
+			       enum pm_device_action action)
+{
+	const struct uart_stm32_config *config = dev->config;
+	struct uart_stm32_data *data = dev->data;
+	int err;
+
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		/* Set pins to active state */
+		err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+		if (err < 0) {
+			return err;
+		}
+
+		/* enable clock */
+		err = clock_control_on(data->clock, (clock_control_subsys_t)&config->pclken[0]);
```

Comment:
It means that UART will always resume with the first clock (`&config->pclken[0]`) on the list.
On the other side, when alternative clock is present, UART is always initialized with alternative clock `&config->pclken[1]`.
It means that clock used for initialization and resume may differ - is it expected behavior?

Reply:
> On the other side, when alternative clock is present, UART is always initialized with alternative clock &config->pclken[1].

**"Alternate" naming is misleading.** 
Sorry for using it initially. "Domain" clock would have been a better choice.
Actually, even when an "domain" clock is provided, both clocks are required for initialization. A "bus" (also called "gating") clock and a fixed clock (/oscillator). W/o bus clock, device cannot operate.
When looking to [RM0434.pdf](https://www.st.com/resource/en/reference_manual/rm0434-multiprotocol-wireless-32bit-mcu-armbased-cortexm4-with-fpu-bluetooth-lowenergy-and-802154-radio-solution-stmicroelectronics.pdf), chapter 33.5.1 "USART block diagram", following description is given:

> The simplified block diagram given in Figure 320 shows two fully-independent clock
> domains:
> - The **usart_pclk** clock domain
>   The **usart_pclk** clock signal feeds the peripheral bus interface. It must be active when
>   accesses to the USART registers are required.
> - The **usart_ker_ck** kernel clock domain.
>   The **usart_ker_ck** is the USART clock source. It is independent from usart_pclk and
>   delivered by the RCC. The USART registers can consequently be written/read even
>   when the usart_ker_ck clock is stopped.
> 
> When the dual clock domain feature is disabled, the **usart_ker_ck** clock is the same as
> the **usart_pclk** clock.

I hope this help clarifying behavior.

**Now, about this code section:**
This code is used only for `PM_DEVICE` modes, so when PM subsystem is aware about devices and is able to suspend/resume them when required.
At suspend time a device PM hook is used to prepare device for LP state:
- clean up pipes, aim is to be ready for next transaction and not generating glitches in following steps:
- set up pins in low consumption mode
- stop clocks
At resumption time, device hook is used to prepare device for operational state:
- set up pins to required function
- enable clocks
Note that enabling/disabling clock is only really useful in `PM_DEVICE_RUNTIME` mode, where PM subsystem can stop/resume device based on its usage, even while the core is still running. In basic `PM_DEVICE` mode, this is indeed superfluous, but makes the code a bit more simple. We could also think to make it conditional to this case.

**Last, why not considering `&config->pclken[1]` in this section ?**
First: Domain clocks may have multiple users (could be even the core itself), and we don't support multiple users awareness on domain clocks. So, we can switch them off blindly.
Second: These clocks are generally taken down in stop modes. See for instance WB Stop 2 mode description:

> [...] In Stop2 mode, all clocks in the V CORE domain are stopped, the PLL, the MSI, the
> HSI16 and the HSE oscillators are disabled. Some peripherals with wakeup capability (I2C3
> and LPUART) can switch on the HSI16 to receive a frame, and switch off the HSI16 after
> receiving the frame if it is not a wakeup frame. In this case the HSI16 clock is propagated
> only to the peripheral requesting it. 

Note that PM susbystem now provides a recent `PM_DEVICE_POWER_DOMAIN` option that I need to look at.

Reply:
Thanks for explanation, now it is clear.
BTW> as usual - good naming makes difference  :-)

Reply:
> BTW> as usual - good naming makes difference :-)

Yup, I'll update naming in tree. That's a bit late, but worth the effort.

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/48679